astropy/time/core.py

```diff
@@ -1002,20 +1035,67 @@ def to_value(self, format, subfmt="*"):
 
             value = tm._shaped_like_input(value)
             cache[key] = value
-        return cache[key]
+        elif self.masked and not isinstance(value, conf._masked_cls):
+            # The requested output masked class can have changed (if the user
+            # changed conf.masked_array_type); if so, change value and cache.
+            value = cache[key] = conf._masked_cls(value)
+
+        return value
 
     @property
     def value(self):
         """Time value(s) in current format."""
         return self.to_value(self.format, None)
 
+    @property
+    def mask(self):
+        if "mask" not in self.cache:
+            mask = getattr(self._time.jd2, "mask", None)
+            if mask is None:
+                mask = np.broadcast_to(np.False_, self._time.jd2.shape)
+            else:
+                # Take a view of any existing mask, so we can set it to readonly.
+                mask = mask.view()
+            mask.flags.writeable = False
+            self.cache["mask"] = mask
+        return self.cache["mask"]
+
     @property
     def masked(self):
-        return self._time.masked
+        return isinstance(self._time.jd1, Masked)
 
     @property
-    def mask(self):
-        return self._time.mask
+    def unmasked(self):
+        """Get an instance without the mask.
+
+        Note that while one gets a new instance, the underlying data will be shared.
+
+        See Also
+        --------
+        astropy.time.Time.filled
+        """
+        return (
+            self._apply(operator.attrgetter("unmasked"))
```

Comment:
Basically, `_apply` applies a function to all array-like attributes and then creates a new instance from the result. Here, the function is to get the `unmasked` attribute.

But looking at this again, this would fail for a location array that is not masked (a possibility I now realize I've happily ignored here - it would work, but the mask would not be reflected in `.mask`... for later??).

To fix this, I've changed this to the following (and added a test):
```
        # Get a new Time instance that has the unmasked versions of all attributes.
        return self._apply(lambda x: getattr(x, "unmasked", x))
```